source/Cosmos.IL2CPU/IL/Conv_I2.cs

```diff
@@ -24,48 +21,45 @@ public override void Execute(_MethodInfo aMethod, ILOpCode aOpCode)
             var xSource = aOpCode.StackPopTypes[0];
             var xSourceSize = SizeOfType(xSource);
             var xSourceIsFloat = TypeIsFloat(xSource);
-            if (xSourceIsFloat)
+
+            if (xSourceSize > 8)
+            {
+                throw new NotImplementedException("Cosmos.IL2CPU.x86->IL->Conv_I2.cs->Error: StackSize > 8 not supported");
+            }
+
+            if (xSourceSize <= 4)
             {
-                if (xSourceSize == 4)
+                if (xSourceIsFloat)
                 {
                     XS.SSE.MoveSS(XMM0, ESP, sourceIsIndirect: true);
                     XS.SSE.ConvertSS2SIAndTruncate(EAX, XMM0);
+                    XS.MoveSignExtend(EAX, AX);
                     XS.Set(ESP, EAX, destinationIsIndirect: true);
                 }
-                else if (xSourceSize == 8)
+                else
+                {
+                    XS.Pop(EAX);
+                    XS.MoveSignExtend(EAX, AX);
+                    XS.Push(EAX);
+                }
+            }
+            else if (xSourceSize <= 8)
+            {
+                if (xSourceIsFloat)
                 {
                     XS.SSE2.MoveSD(XMM0, ESP, sourceIsIndirect: true);
+                    XS.Add(ESP, 4);
                     XS.SSE2.ConvertSD2SIAndTruncate(EAX, XMM0);
                     XS.Set(ESP, EAX, destinationIsIndirect: true);
                 }

```

Comment:
If we had a sort of enum on the internal types it will be yet more better and clear the intent:

```cs
switch (xSourceType)
{
    case xSourceType.Byte:
        XS.Pop(EAX);
        XS.MoveSignExtend(EAX, AL);
        XS.Push(EAX);
        break;
    case 4 when xSourceIsFloat:
        XS.SSE.MoveSS(XMM0, ESP, sourceIsIndirect: true);
        XS.SSE.ConvertSS2SIAndTruncate(EAX, XMM0);
        XS.MoveSignExtend(EAX, AL);
        XS.Set(ESP, EAX, destinationIsIndirect: true);
        break;
    case 8 when !xSourceIsFloat:
        XS.Pop(EAX);
        XS.Add(ESP, 4);
        XS.MoveSignExtend(EAX, AL);
        XS.Push(EAX);
        break;
    case 8 when xSourceIsFloat:
        XS.SSE2.MoveSD(XMM0, ESP, sourceIsIndirect: true);
        XS.Add(ESP, 4);
        XS.SSE2.ConvertSD2SIAndTruncate(EAX, XMM0);
        XS.Set(ESP, EAX, destinationIsIndirect: true);
        break;
    default:
        throw new NotImplementedException($"Cosmos.IL2CPU.x86->IL->Conv_I1.cs->Error: StackSize ${xSourceSize} for ${(xSourceIsFloat ? "float" : "non float")} not supported");
}
```